### ISSUE TITLE
Bug: All child components will rerender if providers not rendering children from props 

### ISSUE THREAD

**Issue by ddhp** [facebook/react#26978]
React version: 18.2.0

## Steps To Reproduce

1. Please check the codesandbox provided below
2. There are 2 components, which are identical except `<Wrapper />` directly renders `<MyPage />` and `LevelContextProvider` renders the same component from `props.children`
3. After clicking `Update context` button, we can see all children in `<Wrapper />` will re-render and `LevelContextProvider` only re-render the ones with `useContext`

Link to code example:
https://codesandbox.io/s/amazing-lewin-wnpndn?file=/src/App.tsx

## The current behavior
as explained above

## The expected behavior
Not sure if current behavior is expected, haven't seem any document about rendering `props.children` in [<Context.Provider />](https://react.dev/reference/react/useContext). It will be much appreciated if someone can provide some information, thanks in advance!

**Comment by markerikson** [comment]
This is all known and expected React behavior:

- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#component-render-optimization-techniques
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#context-updates-and-render-optimizations

**Comment by ddhp** [comment]
Alright, thanks for your time to reply!

tl;dr: 
You probably don't have to re-render components right under context providers, 2 ways to approach:
- React.memo()
- render `props.children` (react skips re-rendering components with the same reference)
